File: crates/process_web/src/api/collect_config.rs
use std::sync::Arc;

use anyhow::Result;
use axum::{
    extract::{Path, State},
    routing::{get, post},
    Json, Router,
};
use serde::Deserialize;
use tracing::debug;
use ts_rs::TS;
use uuid::Uuid;

use crate::api::common::{AppError, AppState, LogTask, PaginationPayload, ResJson, ResJsonWithPagination};
use crate::entity::collect_config::Model;
use crate::service::collect_config_service::CollectConfigService;
use crate::{bool_response, data_response, pagination_response};

pub fn set_routes() -> Router<Arc<AppState>> {
    Router::new()
        .route("/find_by_id/:id", get(find_by_id))
        .route("/list", post(list))
        .route("/add", post(add))
        .route("/update_by_id/:id", post(update_by_id))
        .route("/del/:id", get(del))
        .route("/execute/:id", get(execute))
}

async fn find_by_id(
    state: State<Arc<AppState>>,
    Path(id): Path<i32>,
) -> Result<ResJson<Model>, AppError> {
    let res = CollectConfigService::find_by_id(&state.conn, id).await;

    data_response!(res)
}

#[derive(Deserialize, TS)]
#[ts(
    export,
    export_to = "ui/api/models/auto-generates/CollectConfigListParams.ts",
    rename = "CollectConfigListParams"
)]
pub struct ListParams {
    pub name: Option<String>,
}

async fn list(
    state: State<Arc<AppState>>,
    Json(payload): Json<PaginationPayload<ListParams>>,
) -> Result<ResJsonWithPagination<Model>, AppError> {
    let res = CollectConfigService::list(
        &state.conn,
        payload.current,
        payload.page_size,
        payload.data,
    )
    .await;

    pagination_response!(res, payload.current, payload.page_size)
}
async fn add(
    state: State<Arc<AppState>>,
    Json(payload): Json<Model>,
) -> Result<ResJson<Model>, AppError> {
    let res = CollectConfigService::add(state.0, payload).await;

    data_response!(res)
}

async fn update_by_id(
    state: State<Arc<AppState>>,
    Path(id): Path<i32>,
    Json(payload): Json<Model>,
) -> Result<ResJson<Model>, AppError> {
    let res = CollectConfigService::update_by_id(state.0, id, payload).await;

    data_response!(res)
}

async fn del(state: State<Arc<AppState>>, Path(id): Path<i32>) -> Result<ResJson<bool>, AppError> {
    let res = CollectConfigService::delete(state.0, id).await;

    bool_response!(res)
}

/// 执行id所配置的采集任务
pub async fn execute(
    state: State<Arc<AppState>>,
    Path(id): Path<i32>,
) -> Result<ResJson<bool>, AppError> {
    let data = CollectConfigService::find_by_id(&state.conn, id).await?;

    let task_id = Uuid::new_v4().simple();
    let st = state.clone();
    let t_id = task_id.clone();
    let log_task = LogTask::new();
    let cloned_token = log_task.token.clone();
    let mut task = state.log_task.write().await;
    task.insert(task_id, log_task);
    drop(task);

    tokio::task::spawn(async move {
        tokio::select! {
            _ = cloned_token.cancelled() => {
                // The token was cancelled, task can shut down
                debug!("cloned_token {cloned_token:?}");
            }
            _ = CollectConfigService::execute_task(&st, &data, t_id) => {}
        }
    });
    let res: anyhow::Result<bool> = Ok(true);
    bool_response!(res)
}

File: crates/process_web/ui/api/collect_config.ts
import { CollectConfig } from "@/api/models/CollectConfig";
import {
  http_get,
  http_post,
  PaginationPayload,
  ResJson,
  ResJsonWithPagination,
} from "@/api/common";

export const PREFIX = "/collect_config";

export const FIND_BY_ID = `${PREFIX}/find_by_id/`;

export async function find_by_id(id: number): Promise<ResJson<CollectConfig>> {
  return http_get(`${FIND_BY_ID}${id}`);
}

export const LIST = `${PREFIX}/list/`;
export async function list(
  payload: PaginationPayload<CollectConfig>
): Promise<ResJsonWithPagination<CollectConfig>> {
  return http_post(`${PREFIX}/list`, {
    body: JSON.stringify(payload),
  });
}

export const ADD = `${PREFIX}/add/`;
export async function add(
  payload: CollectConfig
): Promise<ResJson<CollectConfig>> {
  console.log(payload);
  return http_post(`${PREFIX}/add`, {
    body: JSON.stringify(payload),
  });
}

export const UPDATE_BY_ID = `${PREFIX}/update_by_id/`;
export async function update_by_id(
  id: number,
  payload: CollectConfig
): Promise<ResJson<CollectConfig>> {
  return http_post(`${PREFIX}/update_by_id/${id}`, {
    body: JSON.stringify(payload),
  });
}

export const DEL = `${PREFIX}/del/`;
export async function del(id: number): Promise<ResJson<boolean>> {
  return http_get(`${PREFIX}/del/${id}`);
}

export const EXECUTE = `${PREFIX}/execute/`;
export async function execute(id: number): Promise<ResJson<string[]>> {
  return http_get(`${PREFIX}/execute/${id}`);
}
